core/controllers/skill_editor_test.py

```diff
@@ -96,47 +136,142 @@ def test_editable_skill_handler_put(self):
+            # enabled.
+            with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', False):
+                self.delete_json(self.url, expect_errors=True,
+                                 expected_status_int=404)
+        # Check DELETE returns 500 when the skill still has associated
+        # questions.
+        feconf_swap = self.swap(feconf, 'ENABLE_NEW_STRUCTURES', True)
+        skill_has_questions_swap = self.swap(
+            skill_services, 'skill_has_associated_questions', lambda x: True)
+        with feconf_swap, skill_has_questions_swap:
+            self.delete_json(self.url, expect_errors=True,
+                             expected_status_int=500)
+        self.logout()
+
+
+class SkillPublishHandlerTest(BaseSkillEditorControllerTest):
+    """Tests for SkillPublishHandler."""
+
+    def setUp(self):
+        super(SkillPublishHandlerTest, self).setUp()
+        self.url = '%s/%s' % (feconf.SKILL_PUBLISH_URL_PREFIX, self.skill_id)
+
+    def test_skill_publish_handler_succeeds(self):
+        self.login(self.ADMIN_EMAIL)
+        with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', True):
+            # Check that an admin can publish a skill.
+            csrf_token = self._get_csrf_token_for_put()
+            self.put_json(self.url, {'version': 1}, csrf_token=csrf_token)
+        self.logout()
+
+    def test_skill_publish_handler_fails(self):
+
+        def _publish_skill(*_args):
+            raise Exception()
+
+        self.login(self.ADMIN_EMAIL)
+        with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', True):
+            csrf_token = self._get_csrf_token_for_put()
+            # Check that a skill cannot be published when the payload has no
+            # version.
+            self.put_json(self.url, {}, csrf_token=csrf_token,
+                          expect_errors=True, expected_status_int=400)
+            # Check that a skill cannot be published when the payload's version
+            # is different from the skill's version.
+            self.put_json(self.url, {'version': -1}, csrf_token=csrf_token,
+                          expect_errors=True, expected_status_int=400)
+            # Check that a non-existing skill cannot be published.
+            url = '%s/non-existing-id' % (feconf.SKILL_PUBLISH_URL_PREFIX)
+            self.put_json(url, {'version': 1}, csrf_token=csrf_token,
+                          expect_errors=True, expected_status_int=500)
+
+            # Check that the status is 401 when call to publish_skill raises an
+            # exception.
+            skill_services_swap = self.swap(
+                skill_services, 'publish_skill', _publish_skill)
+            with skill_services_swap:
+                csrf_token = self._get_csrf_token_for_put()
+                self.put_json(self.url, {'version': 1}, csrf_token=csrf_token,
+                              expect_errors=True, expected_status_int=401)
+        self.logout()
+
+
+def _get_skill_by_id(*_args, **_kwargs):
```

Comment:
Don't declare these globally -- make them private methods in the class and maybe call them _mock_get_skill_by_id or similar.

Also typically I don't think there are underscores before *args and **kwargs. That said, probably better to use the actual args that match the mocked method. You can prefix each of them by `unused_` to silence lint errors (assuming they are actually unused).

Reply:
Prefixing unused variables only work for positional arguments. For keyword arguments, if we prefix them with `unused`, the signature of the mock method will no long match that of the real method. I guess we have to use something like `**unused_kwargs`.

Reply:
I have made the methods private and prefix `unused` for some arguments in the mock methods. Please take a look.